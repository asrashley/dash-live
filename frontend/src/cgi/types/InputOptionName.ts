export interface InputOptionName {
    shortName: string;
    fullName: string;
    cgiName: string;
}