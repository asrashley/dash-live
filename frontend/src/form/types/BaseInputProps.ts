import type { ReadonlySignal } from "@preact/signals";
import { StaticInputProps } from "./StaticInputProps";

export type BaseInputProps = {
  name: string;
  id: string;
  type: StaticInputProps["type"];
  className: string | ReadonlySignal<string>;
  title: string;
  placeholder?: string;
  "aria-describedby": string;
  disabled: ReadonlySignal<boolean>;
  onInput: (ev: Event) => void;
};

