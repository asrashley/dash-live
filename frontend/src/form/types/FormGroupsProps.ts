import { InputFormGroup } from "./InputFormGroup";
import { FormRowMode } from "./FormRowMode";
import { SetValueFunc } from "./SetValueFunc";
import { InputProps } from "./InputProps";

export interface FormGroupsProps extends Pick<InputProps, "data" | "disabledFields"> {
  groups: InputFormGroup[];
  expand?: string;
  mode: FormRowMode;
  layout?: number[];
  setValue: SetValueFunc;
}
