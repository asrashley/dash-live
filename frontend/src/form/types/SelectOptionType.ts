export interface SelectOptionType {
    name?: string;
    title: string;
    value: string;
}
