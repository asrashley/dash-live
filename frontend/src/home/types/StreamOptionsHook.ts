import { createContext } from "preact";
import { UseStreamOptionsHook } from "../hooks/useStreamOptions";

export const StreamOptionsContext = createContext<UseStreamOptionsHook>(null);
