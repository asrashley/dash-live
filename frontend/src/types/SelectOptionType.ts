export interface SelectOptionType {
    checked?: boolean;
    disabled?: boolean;
    name?: string;
    selected?: boolean;
    title: string;
    value: string;
}
