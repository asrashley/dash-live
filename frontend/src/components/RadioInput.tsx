import { useCallback } from "preact/hooks";
import { SelectOptionType } from "../types/SelectOptionType";

interface RadioOptionProps extends SelectOptionType {
  setValue: (name: string, value: string | number) => void;
}

function RadioOption({ name, selected, title, value, disabled, setValue }: RadioOptionProps) {
  const onClick = useCallback(() => {
    setValue(name, value);
  }, [name, setValue, value]);
  return <div class="form-check">
    <input
      class="form-check-input"
      type="radio"
      name={name}
      id={`radio-${name}-${value}`}
      onClick={onClick}
      value={value}
      checked={selected}
      disabled={disabled}
    />
    <label class="form-check-label" for={`radio-${name}-${value}`}>{title}</label>
  </div>;
}

export interface RadioInputProps {
  options: SelectOptionType[];
  setValue: RadioOptionProps['setValue'];
}

export function RadioInput({ options, setValue }) {
  return <div>
    {options.map(
      (opt) => <RadioOption key={opt.value} setValue={setValue} {...opt} />
    )}
  </div>;
}
