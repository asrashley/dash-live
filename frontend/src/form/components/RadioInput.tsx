import { useComputed, type ReadonlySignal } from "@preact/signals";
import { useCallback } from "preact/hooks";

import { SelectOptionType } from "../types/SelectOptionType";
import { FormGroupsProps } from "../types/FormGroupsProps";

interface RadioOptionProps extends SelectOptionType {
  setValue: (name: string, value: string | number) => void;
  currentValue: ReadonlySignal<string>;
  disabledFields: FormGroupsProps['disabledFields'];
}

function RadioOption({
  name,
  title,
  value,
  currentValue,
  disabledFields,
  setValue,
}: RadioOptionProps) {
  const onClick = useCallback(() => {
    setValue(name, value);
  }, [name, setValue, value]);
  const checked = useComputed<boolean>(() => value === currentValue.value);
  const disabled = useComputed<boolean>(() => !!disabledFields.value[`${name}__${value}`]);
  const id = `radio-${name}-${value}`;

  return (
    <div class="form-check">
      <input
        class="form-check-input"
        type="radio"
        name={name}
        id={id}
        onClick={onClick}
        value={value}
        checked={checked}
        disabled={disabled}
      />
      <label class="form-check-label" for={id}>
        {title}
      </label>
    </div>
  );
}

export interface RadioInputProps {
  name: string;
  options: SelectOptionType[];
  setValue: RadioOptionProps["setValue"];
  value: ReadonlySignal<string>;
  disabledFields: FormGroupsProps['disabledFields'];
}

export function RadioInput({
  name,
  options,
  value,
  disabledFields,
  setValue,
}: RadioInputProps) {
  return (
    <div>
      {options.map((opt) => (
        <RadioOption
          key={opt.value}
          name={name}
          disabledFields={disabledFields}
          setValue={setValue}
          currentValue={value}
          {...opt}
        />
      ))}
    </div>
  );
}
