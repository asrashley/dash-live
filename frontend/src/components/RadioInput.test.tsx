import { afterEach, describe, expect, test, vi } from "vitest";
import { fireEvent } from "@testing-library/preact";

import { renderWithProviders } from "../test/renderWithProviders";
import { RadioInput } from "./RadioInput";
import { SelectOptionType } from "../types/SelectOptionType";

describe("RadioInput", () => {
  const setValue = vi.fn();
  const name = "rtest";
  const options: SelectOptionType[] = [
    {
      name,
      selected: true,
      title: "option one",
      value: "one",
      disabled: false,
    },
    {
      name,
      selected: false,
      title: "option two",
      value: "two",
      disabled: false,
    },
  ];

  afterEach(() => {
    vi.clearAllMocks();
  });

  test("should display radio inputs", () => {
    const { asFragment, getBySelector, getByText } = renderWithProviders(
      <RadioInput options={options} setValue={setValue} />
    );
    getByText(options[0].title);
    getByText(options[1].title);
    const inp = getBySelector(`#radio-${name}-${options[0].value}`) as HTMLInputElement;
    expect(inp.checked).toEqual(true);
    expect(asFragment()).toMatchSnapshot();
  });

  test("can set value", () => {
    const { getBySelector } = renderWithProviders(
      <RadioInput options={options} setValue={setValue} />
    );
    const inp = getBySelector(`#radio-${name}-${options[1].value}`) as HTMLInputElement;
    expect(inp.disabled).toEqual(false);
    fireEvent.click(inp);
    expect(setValue).toHaveBeenCalledTimes(1);
    expect(setValue).toHaveBeenCalledWith(name, options[1].value);
  });

  test("can disable option", () => {
    const opts = [ options[0], {
        ...options[1],
        disabled: true,
    }];
    const { getBySelector } = renderWithProviders(
      <RadioInput options={opts} setValue={setValue} />
    );
    const inp = getBySelector(`#radio-${name}-${options[1].value}`) as HTMLInputElement;
    expect(inp.disabled).toEqual(true);
  });
});
