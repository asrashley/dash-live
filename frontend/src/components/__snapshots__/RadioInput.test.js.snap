// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RadioInput > should display radio inputs 1`] = `
<DocumentFragment>
  <div>
    <div
      class="form-check"
    >
      <input
        class="form-check-input"
        id="radio-rtest-one"
        name="rtest"
        type="radio"
        value="one"
      />
      <label
        class="form-check-label"
        for="radio-rtest-one"
      >
        option one
      </label>
    </div>
    <div
      class="form-check"
    >
      <input
        class="form-check-input"
        id="radio-rtest-two"
        name="rtest"
        type="radio"
        value="two"
      />
      <label
        class="form-check-label"
        for="radio-rtest-two"
      >
        option two
      </label>
    </div>
  </div>
</DocumentFragment>
`;
